synapse/api/ratelimiting.py

```diff
@@ -16,75 +17,153 @@
 from typing import Any, Optional, Tuple
 
 from synapse.api.errors import LimitExceededError
+from synapse.util import Clock
 
 
 class Ratelimiter(object):
     """
-    Ratelimit message sending by user.
+    Ratelimit actions marked by arbitrary keys.
+
+    Args:
+        clock: A homeserver clock, for retrieving the current time
+        rate_hz: The long term number of actions that can be performed in a second.
+        burst_count: How many actions that can be performed before being limited.
     """
 
-    def __init__(self):
-        self.message_counts = (
-            OrderedDict()
-        )  # type: OrderedDict[Any, Tuple[float, int, Optional[float]]]
+    def __init__(self, clock: Clock, rate_hz: float, burst_count: int):
+        self.clock = clock
+        self.rate_hz = rate_hz
+        self.burst_count = burst_count
+
+        # A ordered dictionary keeping track of actions, when they were last
+        # performed and how often. Each entry is a mapping from a key of arbitrary type
+        # to a tuple representing:
+        #   * How many times an action has occurred since a point in time
+        #   * The point in time
+        #   * The rate_hz of this particular entry. This can vary per request
+        self.actions = OrderedDict()  # type: OrderedDict[Any, Tuple[float, int, float]]
 
-    def can_do_action(self, key, time_now_s, rate_hz, burst_count, update=True):
+    def can_do_action(
+        self,
+        key: Any,
+        time_now_s: Optional[int] = None,
+        update: bool = True,
```

Comment:
It is a little weird to have a non-optional parameter after an optional one?

Reply:
Well, `update` is optional in the sense that the calling function doesn't have to provide it, but yes it can't be `None`. Did you mean optional as in it cannot be `None`?

Reply:
Yes, I meant in terms of it not being `None`. A call like this is kind of weird:
```python
can_do_action('foo', None, False)
```

I don't think it is a big deal though.

Reply:
Actually, is `time_now_s` *only* used in tests?

It seems weird to me that you can provide a `Clock`, but also can provide `time_now_s`, it seems like these would be two different ways of allowing for tests to control time.

If we need both I'd suggest renaming `time_now_s` to `_time_now_s` and putting it as the last kwarg of `can_do_action`. This should make it more obvious that it isn't meant to be used in "normal" usage.

Reply:
Good idea! Yes, it seems so and I mentioned it in the docstring. Good point about the placement and renaming though. I don't like the pattern and believe code shouldn't be written for tests, but I think it'd be much more complicated to write tests that messed with the clock object.